[css-masking] Implement `mask-border` properties

nt1m opened the pull request:
#### ecf24fddc3cb9df61b9c8cb1fb9e9f5efaee267f
<pre>
[css-masking] Implement `mask-border` properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=229900">https://bugs.webkit.org/show_bug.cgi?id=229900</a>
rdar://82991599

Reviewed by Darin Adler.

This patch aliases -webkit-mask-box-image-* to mask-border-*, with a few changes:
- the initial value of mask-border-slice is now `0` instead of `0 fill`
- the `-webkit-mask-box-image` prop is implemented as a separate shorthand that sets the `fill` on the slice property unlike the `mask-border` shorthand

This is similar to how -webkit-border-image and border-image are setup.

Setting the mask shorthand (but not -webkit-mask) now also resets the `mask-border-*` properties as specified.

Also mass rename maskBoxImage in the codebase to maskBorder.

Specification at: <a href="https://drafts.fxtf.org/css-masking/#the-mask-border">https://drafts.fxtf.org/css-masking/#the-mask-border</a>

* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-webkit-mask-box-image.html:
* LayoutTests/fast/css/mask-box-image-parsing-expected.txt:
* LayoutTests/fast/css/mask-box-image-parsing.html:
* LayoutTests/fast/css/uri-token-parsing-expected.txt:
* LayoutTests/fast/css/uri-token-parsing.html:
* LayoutTests/fast/masking/parsing-mask-repeat-expected.txt:
* LayoutTests/fast/masking/parsing-mask-repeat.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/inheritance.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/mask-shorthand-subproperties-reset-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/mask-shorthand-subproperties-reset.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeBorderImage const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::isInitialValueForLonghand):
(WebCore::initialValueTextForLonghand):
(WebCore::initialValueIDForLonghand):
(WebCore::CSSPropertyParser::consumeBorderImage):
(WebCore::CSSPropertyParser::consumeBackgroundShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBorderImageComponents):
* Source/WebCore/rendering/InlineBoxPainter.cpp:
(WebCore::InlineBoxPainter::paintMask):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::paintMaskImages):
(WebCore::RenderBox::maskClipRect):
(WebCore::RenderBox::imageChanged):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleDidChange):
(WebCore::RenderElement::willBeDestroyed):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::defaultMaskData):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareDataChangeRequiresLayerRepaint):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialMaskBorderSource):
(WebCore::RenderStyle::initialMaskBoxImageSource): Deleted.
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasMask const):
(WebCore::RenderStyle::maskBorder const):
(WebCore::RenderStyle::maskBorderOutsets const):
(WebCore::RenderStyle::maskBorderSource const):
(WebCore::RenderStyle::maskBoxImage const): Deleted.
(WebCore::RenderStyle::maskBoxImageOutsets const): Deleted.
(WebCore::RenderStyle::maskBoxImageSource const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setMaskBorder):
(WebCore::RenderStyle::setMaskBorderSource):
(WebCore::RenderStyle::setMaskBoxImage): Deleted.
(WebCore::RenderStyle::setMaskBoxImageSource): Deleted.
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleReflection.h:
(WebCore::StyleReflection::StyleReflection):
* Source/WebCore/rendering/style/WillChangeData.cpp:
(WebCore::WillChangeData::propertyCreatesStackingContext):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertReflection):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::ApplyPropertyBorderImageModifier::applyInitialValue):
(WebCore::Style::ApplyPropertyBorderImageModifier::getValue):
(WebCore::Style::ApplyPropertyBorderImageModifier::setValue):
* Source/WebCore/style/StylePendingResources.cpp:
(WebCore::Style::loadPendingResources):

Canonical link: <a href="https://commits.webkit.org/267460@main">https://commits.webkit.org/267460@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79f7ffe331f703af216d41ab7396eafb1bc67718

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17387 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18400 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15586 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20171 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17082 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17909 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16816 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14386 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19183 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14460 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15070 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15447 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15236 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19544 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15831 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13449 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14910 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/3992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19397 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2052 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15656 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -4143,12 +4144,12 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
-        case CSSPropertyWebkitMaskBoxImageSlice:
-        case CSSPropertyWebkitMaskBoxImageSource:
-        case CSSPropertyWebkitMaskBoxImageWidth:
+        case CSSPropertyMaskBorder:
```
I think this was sorted in alphabetical order before.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -940,7 +940,7 @@ static constexpr InitialValue initialValueForLonghand(CSSPropertyID longhand)
     case CSSPropertyTextUnderlinePosition:
     case CSSPropertyTop:
-    case CSSPropertyWebkitMaskBoxImageWidth:
+    case CSSPropertyMaskBorderWidth:
```
I think this was sorted in alphabetical order before.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1056,7 +1056,7 @@ static constexpr InitialValue initialValueForLonghand(CSSPropertyID longhand)
     case CSSPropertyTransform:
     case CSSPropertyTranslate:
-    case CSSPropertyWebkitMaskBoxImageSource:
+    case CSSPropertyMaskBorderSource:
```
I think this was sorted in alphabetical order before.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1129,13 +1129,15 @@ static constexpr InitialValue initialValueForLonghand(CSSPropertyID longhand)
         return CSSValueStretch;
     case CSSPropertyBorderImageSlice:
         return InitialNumericValue { 100, CSSUnitType::CSS_PERCENTAGE };
+    case CSSPropertyMaskBorderSlice:
```
I think this was sorted in alphabetical order before.

darinadler commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -118,10 +118,10 @@ class BuilderCustom {
-    DECLARE_PROPERTY_CUSTOM_HANDLERS(WebkitMaskBoxImageRepeat);
-    DECLARE_PROPERTY_CUSTOM_HANDLERS(WebkitMaskBoxImageSlice);
-    DECLARE_PROPERTY_CUSTOM_HANDLERS(WebkitMaskBoxImageWidth);
+    DECLARE_PROPERTY_CUSTOM_HANDLERS(MaskBorderOutset);
```
I think this was sorted in alphabetical order before.

nt1m committed:
[css-masking] Implement `mask-border` properties
https://bugs.webkit.org/show_bug.cgi?id=229900
rdar://82991599

Reviewed by Darin Adler.

This patch aliases -webkit-mask-box-image-* to mask-border-*, with a few changes:
- the initial value of mask-border-slice is now `0` instead of `0 fill`
- the `-webkit-mask-box-image` prop is implemented as a separate shorthand that sets the `fill` on the slice property unlike the `mask-border` shorthand

This is similar to how -webkit-border-image and border-image are setup.

Setting the mask shorthand (but not -webkit-mask) now also resets the `mask-border-*` properties as specified.

Also mass rename maskBoxImage in the codebase to maskBorder.

Specification at: https://drafts.fxtf.org/css-masking/#the-mask-border

* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-webkit-mask-box-image.html:
* LayoutTests/fast/css/mask-box-image-parsing-expected.txt:
* LayoutTests/fast/css/mask-box-image-parsing.html:
* LayoutTests/fast/css/uri-token-parsing-expected.txt:
* LayoutTests/fast/css/uri-token-parsing.html:
* LayoutTests/fast/masking/parsing-mask-repeat-expected.txt:
* LayoutTests/fast/masking/parsing-mask-repeat.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/inheritance.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/mask-shorthand-subproperties-reset-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/mask-shorthand-subproperties-reset.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeBorderImage const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::isInitialValueForLonghand):
(WebCore::initialValueTextForLonghand):
(WebCore::initialValueIDForLonghand):
(WebCore::CSSPropertyParser::consumeBorderImage):
(WebCore::CSSPropertyParser::consumeBackgroundShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBorderImageComponents):
* Source/WebCore/rendering/InlineBoxPainter.cpp:
(WebCore::InlineBoxPainter::paintMask):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::paintMaskImages):
(WebCore::RenderBox::maskClipRect):
(WebCore::RenderBox::imageChanged):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleDidChange):
(WebCore::RenderElement::willBeDestroyed):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::defaultMaskData):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareDataChangeRequiresLayerRepaint):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialMaskBorderSource):
(WebCore::RenderStyle::initialMaskBoxImageSource): Deleted.
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasMask const):
(WebCore::RenderStyle::maskBorder const):
(WebCore::RenderStyle::maskBorderOutsets const):
(WebCore::RenderStyle::maskBorderSource const):
(WebCore::RenderStyle::maskBoxImage const): Deleted.
(WebCore::RenderStyle::maskBoxImageOutsets const): Deleted.
(WebCore::RenderStyle::maskBoxImageSource const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setMaskBorder):
(WebCore::RenderStyle::setMaskBorderSource):
(WebCore::RenderStyle::setMaskBoxImage): Deleted.
(WebCore::RenderStyle::setMaskBoxImageSource): Deleted.
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleReflection.h:
(WebCore::StyleReflection::StyleReflection):
* Source/WebCore/rendering/style/WillChangeData.cpp:
(WebCore::WillChangeData::propertyCreatesStackingContext):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertReflection):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::ApplyPropertyBorderImageModifier::applyInitialValue):
(WebCore::Style::ApplyPropertyBorderImageModifier::getValue):
(WebCore::Style::ApplyPropertyBorderImageModifier::setValue):
* Source/WebCore/style/StylePendingResources.cpp:
(WebCore::Style::loadPendingResources):

Canonical link: https://commits.webkit.org/267460@main
